File: day1/puzzle2.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const char *digits[] = {
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
    "zero",
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
};

int main(int argc, char *argv[]) {
    FILE *input = fopen("puzzle_input.txt", "rt");
    if (input == NULL) {
        return EXIT_FAILURE;
    }
    char line[256] = {0};
    int sum = 0;
    while (fgets(line, sizeof(line), input) != NULL) {
        int first = 0;
        int first_index = strlen(line);
        int last = 0;
        int last_index = -1;
        for (int i = 0; i < 20; i++) {
            if ((i == 0) || (i == 10)) {
                continue;
            }
            int start_index = -1;
            while (true) {
                char *p = strstr(line + start_index + 1, digits[i]);
                if (p == NULL) {
                    break;
                }
                start_index = p - line;
                if (start_index < first_index) {
                    first_index = start_index;
                    first = i % 10;
                }
                if (start_index > last_index) {
                    last_index = start_index;
                    last = i % 10;
                }
            }
        }
        sum += first * 10 + last;
    }
    fclose(input);
    printf("sum = %d\n", sum);
    return EXIT_SUCCESS;
}

File: day1/puzzle2.f90
program puzzle2
    implicit none

    integer :: io, sum = 0
    integer :: index1, index2, len, first_index, first, last_index, last
    character(len=80) :: buffer
    character(len=5), dimension(20) :: digits

    digits = [character(len=5) :: "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", &
              "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

    open(1, file="puzzle_input.txt", status="old", action="read", iostat=io)
    if (io /= 0) then
        print '("error opening file: ",i0)', io
        call exit(io)
    end if
    do while (io == 0)
        read(1, '(a)', iostat=io) buffer
        if (io /= 0) exit
        first_index = 80
        last_index = 0
        do index1=1,len_trim(buffer)
            do index2=1,20
                len = len_trim(digits(index2))
                if (buffer(index1:index1+len-1) == digits(index2)(1:len)) then
                    if (index1 < first_index) then
                        first_index = index1
                        first = mod(index2-1, 10)
                    end if
                    if (index1 > last_index) then
                        last_index = index1
                        last = mod(index2-1, 10)
                    end if
                end if
            end do
        end do
        sum = sum + first * 10 + last
    end do
    close(1)
    print '("sum = ",i0)', sum
end program puzzle2
